src/Chapter3.hs

```diff
@@ -1006,6 +1119,132 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+data KnightAction = AttackAction | PotionAction Health | SpellAction Defense deriving (Show, Eq)
+data MonsterAction = HitAction | RunAction deriving (Show, Eq)
+
+newtype Health = Health { unHealth :: Int } deriving (Show, Eq)
+newtype Attack = Attack { unAttack :: Int } deriving (Show, Eq)
+newtype Defense = Defense { unDefense :: Int } deriving (Show, Eq)
+
+instance Append Health where
+  append (Health h1) (Health h2) = Health (h1 + h2)
+
+instance Append Defense where
+  append (Defense d1) (Defense d2) = Defense (d1 + d2)
```

Comment:
Nice way of reusing the `Append` typeclass! If you are interested, this typeclass is known as `Semigroup` in the Haskel world:

* http://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Semigroup.html